ister_gui.py

```diff
@@ -644,6 +672,247 @@ def validate_condition(self):
         return True
 
 
+class ChooseAction(ProcessStep):
+    """UI to select installation path (Install, Shell, or Repair)"""
+    def __init__(self):
+        super(ChooseAction, self).__init__()
+        self.choices = {
+            'Install': 'Install          - Install Clear Linux OS',
+            'Repair': 'Repair           - Attempt to repair host OS',
+            'Shell': 'Shell            - Open shell in host OS',
+            'Update Installer': 'Update Installer - Update installer to '
+                                'latest version and restart the installer'
+        }
+        self._clicked = None
+        self.error = False
+
+    def build_ui_widgets(self):
+        self._ui_widgets = []
+        for key in sorted(self.choices):
+            button = urwid.Button(self.choices[key])
+            urwid.connect_signal(button, 'click', self._item_chosen, key)
+            button = urwid.AttrMap(button, None, focus_map='reversed')
+            self._ui_widgets.append(button)
+
+    def _item_chosen(self, _, choice):
+        self._clicked = choice
+        raise urwid.ExitMainLoop()
+
+    def build_ui(self):
+        self._ui = SimpleForm('Choose Action', self._ui_widgets,
+                              buttons=['Previous'])
+
+    def run_ui(self):
+        return self._ui.do_form()
+
+    def handler(self, config):
+        if not self._ui_widgets:
+            self.build_ui_widgets()
+
+        if not self._ui:
+            self.build_ui()
+
+        self._clicked = ''
+
+        if self._ui:
+            self._action = self.run_ui()
+
+        if self._clicked:
+            self._action = self._clicked
+
+        if self._action == 'Shell':
+            self.chroot_shell(config)
+
+        if self._action == 'Repair':
+            self.repair_host(config)
+
+        if self._action == 'Update Installer':
+            self.update_installer(config)
+
+        return self._action
+
+    def chroot_shell(self, config):
+        """Chroot into host os (if one is found) and provide a shell"""
+        self.error = False
```

Comment:
Hrm do you ignore some things setting self.error? If you do then setting this before testing makes sense. Otherwise I'd say any path that acts on a self.error should also reset it after it checks.

Prevents cases where:

self.error = False
self.do_one()
if self.error:
 self.do_two()
else:
 self.do_three()
if self.error:
  # are we sure do_two and do_three manage self.error or did only do_two set it?